netlib/http/http2/connections.py

```diff
@@ -1,420 +1,285 @@
-        CANCEL=0x8,
-        COMPRESSION_ERROR=0x9,
-        CONNECT_ERROR=0xa,
-        ENHANCE_YOUR_CALM=0xb,
-        INADEQUATE_SECURITY=0xc,
-        HTTP_1_1_REQUIRED=0xd
-    )
 
-    CLIENT_CONNECTION_PREFACE = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"
-
-    def __init__(
-        self,
-        tcp_handler=None,
-        rfile=None,
-        wfile=None,
-        is_server=False,
-        dump_frames=False,
-        encoder=None,
-        decoder=None,
-        unhandled_frame_cb=None,
-    ):
-        self.tcp_handler = tcp_handler or TCPHandler(rfile, wfile)
-        self.is_server = is_server
-        self.dump_frames = dump_frames
-        self.encoder = encoder or Encoder()
-        self.decoder = decoder or Decoder()
-        self.unhandled_frame_cb = unhandled_frame_cb
-
-        self.http2_settings = frame.HTTP2_DEFAULT_SETTINGS.copy()
-        self.current_stream_id = None
-        self.connection_preface_performed = False
-
-    def read_request(
-        self,
-        __rfile,
-        include_body=True,
-        body_size_limit=None,
-        allow_empty=False,
-    ):
-        if body_size_limit is not None:
-            raise NotImplementedError()
-
-        self.perform_connection_preface()
-
-        timestamp_start = time.time()
-        if hasattr(self.tcp_handler.rfile, "reset_timestamps"):
-            self.tcp_handler.rfile.reset_timestamps()
-
-        stream_id, headers, body = self._receive_transmission(
-            include_body=include_body,
+from ...exceptions import Http2Exception
+from .. import Headers, Request, Response
+from .frame import (
+    Frame, WindowUpdateFrame, SettingsFrame, ContinuationFrame, HeadersFrame, DataFrame
+)
+
+CLIENT_CONNECTION_PREFACE = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"
+
+
+class Http2Connection(object):
+    def __init__(self, connection):
+        self.encoder = Encoder()
+        self.decoder = Decoder()
+        self.lock = threading.RLock()
+
+        self._connection = connection
+
+        self.initiated_streams = 0
+        self.streams = {}
+
+    def read_frame(self):
+        with self.lock:
+            return Frame.from_file(self.rfile)  # TODO: max_body_size
+
+    def send_frame(self, *frames):
+        """
+        Should only be called with multiple frames that MUST be sent in sequence,
+        e.g. a HEADEDRS frame and its CONTINUATION frames
+        """
+        with self.lock:
+            for frame in frames:
+                print("send frame", self.__class__.__name__, repr(frame))
+                d = frame.to_bytes()
+                self._connection.wfile.write(d)
+                self._connection.wfile.flush()
+
+    # FIXME: We should respect the actual connection settings
+    max_frame_size = 2 ** 14
+
+    def send_headers(self, headers, stream_id, end_stream=False):
+        with self.lock:
+            if stream_id is None:
+                stream_id = self._next_stream_id()
+            raw_headers = self.encoder.encode(headers.fields)
```

Comment:
I think we probably want to do it here because we want non-lowercase headers for pathod, but the question whether we want to do this in `make_request` and `make_response` for mitmproxy is a valid one. I think it'd be a good thing in terms of robustness, but then again this creates situations where connections fail without mitmproxy but work once mitmproxy is on the wire - that't not nice either.